allennlp/tests/modules/token_embedders/pretrained_transformer_embedder_test.py

```diff
@@ -69,3 +71,124 @@ def test_end_to_end(self):
+
+        vocab = Vocabulary()
+
+        params = Params(
+            {
+                "token_embedders": {
+                    "bert": {
+                        "type": "pretrained_transformer",
+                        "model_name": "bert-base-uncased",
+                        "max_length": 4,
+                    }
+                }
+            }
+        )
+        token_embedder = BasicTextFieldEmbedder.from_params(vocab=vocab, params=params)
+
+        instance1 = Instance({"tokens": TextField(tokens1, {"bert": token_indexer})})
+        instance2 = Instance({"tokens": TextField(tokens2, {"bert": token_indexer})})
+
+        batch = Batch([instance1, instance2])
+        batch.index_instances(vocab)
+
+        padding_lengths = batch.get_padding_lengths()
+        tensor_dict = batch.as_tensor_dict(padding_lengths)
+        tokens = tensor_dict["tokens"]
+        max_length = max(len(tokens1), len(tokens2))
+
+        # Adds n_segments * 2 special tokens
+        segment_concat_length = int(math.ceil(max_length / 4)) * 2 + max_length
+        assert tokens["bert"]["token_ids"].shape == (2, segment_concat_length)
+
+        assert tokens["bert"]["mask"].tolist() == [
+            [1, 1, 1, 1, 1, 1, 1, 1, 1],
+            [1, 1, 1, 1, 1, 1, 1, 0, 0],
+        ]
+        assert tokens["bert"]["segment_concat_mask"].tolist() == [
+            [1] * segment_concat_length,
+            [1] * (segment_concat_length - 4) + [0] * 4,  # 4 is hard-coded length difference
+        ]
+
+        # Attention mask
+        bert_vectors = token_embedder(tokens)
+        assert bert_vectors.size() == (2, 9, 768)
+
+    def test_fold_long_sequences(self):
+        # Let's just say [PAD] is 0, [CLS] is 1, and [SEP] is 2
+        token_ids = torch.LongTensor(
+            [
+                [1, 101, 102, 103, 104, 2, 1, 105, 106, 107, 108, 2, 1, 109, 2],
+                [1, 201, 202, 203, 204, 2, 1, 205, 206, 207, 208, 2, 0, 0, 0],
+                [1, 301, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+            ]
+        )  # Shape: [3, 15]
+        segment_concat_mask = (token_ids > 0).long()
+
+        folded_token_ids = torch.LongTensor(
+            [
+                [1, 101, 102, 103, 104, 2],
+                [1, 105, 106, 107, 108, 2],
+                [1, 109, 2, 0, 0, 0],
+                [1, 201, 202, 203, 204, 2],
+                [1, 205, 206, 207, 208, 2],
+                [0, 0, 0, 0, 0, 0],
+                [1, 301, 2, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0],
+            ]
+        )
+        folded_segment_concat_mask = (folded_token_ids > 0).long()
+
+        token_embedder = PretrainedTransformerEmbedder("bert-base-uncased", max_length=6)
+
+        folded_token_ids_out, folded_segment_concat_mask_out = token_embedder._fold_long_sequences(
+            token_ids, segment_concat_mask
+        )
+        assert (folded_token_ids_out == folded_token_ids).all()
```

Comment:
We'll often use `numpy.testing.assert_array_equal`. There's also `numpy.testing.assert_array_almost_equal` for cases where floating point math is involved.

Reply:
These are ints, so it's fine. I'm using regular assert here because people were using these in this file before me, and I want to be consistent. If we don't like these in the future, doing a sed wouldn't be hard.

Reply:
👍 for consistency.